source/dealii/dealii_hierarchy_helpers.cc

```diff
@@ -223,13 +232,28 @@ DealIIHierarchyHelpers<dim, VectorType>::build_restrictor(
     }
 
     // Fill delta_correction_matrix
-    for (auto const &entry : delta_correction_acc)
-    {
-      delta_correction_matrix.set(entry.first.first, entry.first.second,
-                                  entry.second);
-    }
     delta_correction_matrix.compress(dealii::VectorOperation::insert);
 
+    auto const range_start = eigenvector_matrix->local_range().first;
+    auto const range_end = eigenvector_matrix->local_range().second;
+
+    dealii::IndexSet locally_owned_elements(eigenvector_matrix->m());
+    locally_owned_elements.add_range(range_start, range_end);
+    dealii::TrilinosWrappers::MPI::Vector vector_eigenvalues(
+        locally_owned_elements, comm);
+    std::vector<unsigned int> local_indices(eigenvalues.size());
+    std::iota(local_indices.begin(), local_indices.end(), range_start);
+    vector_eigenvalues.set(local_indices, eigenvalues);
+    vector_eigenvalues.compress(dealii::VectorOperation::insert);
+
+    Epetra_MultiVector &dummy_multi_vector =
+        vector_eigenvalues.trilinos_vector();
+    auto &dummy_matrix =
+        const_cast<Epetra_CrsMatrix &>(eigenvector_matrix->trilinos_matrix());
+    dummy_matrix.LeftScale(*dummy_multi_vector(0));
```

Comment:
This is good but we need to add a comment building the matrix this way speeds up the computation so that we don't rewrite it later to simplify the code.